sdk/core/azure-core/src/main/java/com/azure/android/core/util/paging/PagedCollection.java

```diff
@@ -0,0 +1,119 @@
+package com.azure.android.core.util.paging;
+
+import androidx.annotation.NonNull;
+
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.Iterator;
+import java.util.LinkedHashMap;
+import java.util.List;
+
+public class PagedCollection<T, P extends Page<T>> implements Iterable<T> {
+    private static final int DEFAULT_PAGE_SIZE = 25; // Is this a good default size?
+
+    private PageRetriever<T, P> pageRetriever;
+    private Integer pageSize;
+    private List<T> items;
+    private LinkedHashMap<String, P> pages;
+    private String nextPageId;
+
+    public PagedCollection(PageRetriever<T, P> pageRetriever) {
+        this.pageSize = DEFAULT_PAGE_SIZE;
+        this.pageRetriever = pageRetriever;
+        items = new ArrayList<>();
+        pages = new LinkedHashMap<>();
+    }
+
+    public PagedCollection(PageRetriever<T, P> pageRetriever, Integer pageSize) {
+        this(pageRetriever);
+
+        if (pageSize == null || pageSize <= 0) {
+            throw new IllegalArgumentException("Page size must not be null and must be greater than zero.");
+        }
+
+        this.pageSize = pageSize;
+    }
+
+    public List<T> getItems() {
+        return items;
+    }
+
+    public Collection<P> getPages() {
+        return pages.values();
+    }
+
+    // A null page Id gets the first page from the service. It will get cached with key: null.
+    public void getPage(String pageId, PagingCallback<T, P> callback) {
+        P page = pages.get(pageId);
```

Comment:
This function seems to indicate a random access cache by pageId but the iterator pattern seems to indicate a sequential access cache.

Reply:
I thought we could allow both types of interaction in the same class. Based on your previous comments it seems like you think that',s not a good idea or is confusing. Let's discuss this on Monday :)